### DISCUSSION TITLE
Embedding in Umbraco

### DISCUSSION THREAD

**Discussion by mattbrailsford** [elsa-workflows/elsa-core#2798] [General]
Apologies if this discussion is a little vague but I'm looking for some input on how one might go about embedding ELSA within an existing application, in this case Umbraco.

I'm an Umbraco developer and I think it could be quite a good fit to have ELSA as an embeddable workflow engine to do advanced approval flows within the CMS and some of it's add-ons (I develop an eCommerce plugin for Umbraco and having ELSA workflows that could be triggered by orders would be super cool).

I've managed to get the basic example working, but all to demos really kind of assume that ELSA is core to the application, but being embedded in Umbraco would mean that Umbraco becomes responsible for a lot of things, specifically around routing and permissions etc, and so I need to get ELSA to install around this.

A few initial things I'm wondering are

1. Can the API endpoints be made to hang off of a prefix so that they can easily be segmented away from the other routing that Umbraco will be responsible for (Maybe ELSA Server could be registered as an Area?)
2. If the workflow designer is made to only be accessible within Umbraco's back office, how would one go about adding authentication requirements just to the ELSA endpoints
3. Looking in the registration code for the endpoints / http activities, I've kinda got the feeling there might be some conflicts that will likely occur here as both systems attempt to configure routing how they want (forced lower case routes, configuring JSON.NET etc)

Has anyone attempted to embed ELSA in Umbraco? or some other existing application that already has fairly advanced routing in place? I'd love to know how best to tackle this.

**Comment by mattbrailsford** [comment]
I've just noticed the https://github.com/elsa-workflows/elsa-orchard-core repository which I guess would be a good starting point to integrate in a similar way with Umbraco.

Can anyone shed any light on the status of that repo? Does it work in Orchard? What is missing / outstanding? And is there a summary of what that project is doing in comparison to the ELSA Server implementation of core?

**Reply by mattbrailsford** [reply]
Ok, so looking through the source of the Orchard Core project it looks like it's only currently capable of being a client to a remote ELSA Server, rather than the Orchard instance being a local server itself.

Would love to know what the steps would be to get the local server instance working.

**Comment by mattbrailsford** [comment]
Regarding item 1 in my original list, I've had some luck with what's written in this article to move all the server API endpoints to `/elsa/api` https://www.strathweb.com/2016/06/global-route-prefix-with-asp-net-core-mvc-revisited/

**Comment by sfmskywalker** [comment]
What you're doing for Umbraco is what I started out doing for Orchard Core. You're right, currently it is merely a client to a remote Elsa server. The next steps are to enable the Orchard application itself to act as a workflow server.

> 1. Can the API endpoints be made to hang off of a prefix so that they can easily be segmented away from the other routing that Umbraco will be responsible for (Maybe ELSA Server could be registered as an Area?)

I don't know, this is something I need to look into. Maybe you're right and `Elsa.Server.Api` could be registered as an area.

> 2. If the workflow designer is made to only be accessible within Umbraco's back office, how would one go about adding authentication requirements just to the ELSA endpoints

That's the key question: how to add authentication requirements **just** to the Elsa endpoints. Currently, I only know how to add this to ALL endpoints from startup, like this:

```csharp
endpoints.MapControllers().RequireAuthorization()
```

But that's not ideal. Perhaps the Elsa API controllers should be annotated with `AuthorizeAttribute` and some custom policy name called e.g. "ElsaPolicy". Then the hosting app can configure this policy as necessary.

>3. Looking in the registration code for the endpoints / http activities, I've kinda got the feeling there might be some conflicts that will likely occur here as both systems attempt to configure routing how they want (forced lower case routes, configuring JSON.NET etc)

For the **HTTP Endpoint** activity, there shouldn't be any conflicts (at least none that I'm aware of) since it does not *yet) integrate with ASP.NET's routing mechanism. It simply relies on a custom middleware (`HttpEndpointMiddleware`) that performs some basic path matching.

But the `Elsa.Activities.Http` library does expose a couple of API controllers (like `Elsa.Server.Api`), so whatever solutions are found for `Elsa.Server.Api` might apply here as well.

Ultimately what I think we need is a clean way to package up API endpoints in such a way that the hosting app has control over these aspects:

- Authorization
- Routing (at least prefix)

For Elsa 3, I am experimenting with Minimal APIs, where the hosting app has precise control over what routes they wish to map to functions exposed by Elsa. So far, it seems to work out nicely, but I'm a little wary about missing out some of the more fancier aspects provided by full-blown controllers. Additionally, the application shouldn't be in control over the complete route without also requiring the client app (the designer) to be configured with each API endpoint. Probably just allowing the application to configure a route prefix would suffice.
